tapeagents/llms.py

```diff
@@ -637,6 +651,84 @@ def _generate(self, prompt: Prompt) -> Generator[LLMEvent, None, None]:
+        }
+        if self.collect_logprobs:
+            data.update(
+                {
+                    "logprobs": 1,
+                    "include_stop_str_in_output": True,
+                    "skip_special_tokens": False,
+                }
+            )
+        base_url = self.base_url if isinstance(self.base_url, str) else random.choice(self.base_url)
+        logger.debug(f"POST request to {base_url}/v1/completions")
+        start_send_request = time.time()
+        r = requests.post(
+            url=f"{base_url}/v1/completions",
+            json=data | self.parameters,
+            headers=headers,
+            stream=self.stream,
+            verify=False,
+        )
+        self._stats["time_send_request"].append(time.time() - start_send_request)
+        if not r.ok:
+            logger.error(f"Failed to get completion: {r.text}")
+            r.raise_for_status()
+        data = r.json()
+        result = []
+        start_postprocess_time = time.time()
+        for i in range(len(prompts)):
+            try:
+                content = data["choices"][i]["text"]
+                if not content:
+                    logger.warning(f"Empty completion {data}")
+
+                logprobs = None
+                if self.collect_logprobs:
+                    completion_logprobs = data["choices"][i]["logprobs"]
+                    # /v1/completions returns logprobs in a format different to /v1/chat/completions
+                    # Before calling self.process_logprobs, we need to convert the logprobs to a
+                    # list of dicts format similar to /v1/chat/completions
+                    
+                    chat_completion_logprobs = [
+                        {"token": completion_logprobs["tokens"][j], "logprob": completion_logprobs["token_logprobs"][j]}
+                        for j in range(len(completion_logprobs["tokens"]))
+                    ]
+                    logprobs = self.make_llm_call_logprobs(prompt_token_ids[i], chat_completion_logprobs)
+                    # <end_of_turn> is the end of message for Gemma2B, eos_token is wrong for this model
+                    for eos_str in [self.tokenizer.eos_token, "<end_of_turn>"]:
+                        if content.endswith(eos_str):
+                            # the eos was added in the case where self.collect_logprobs is True
+                            # TapeAgents is not expecting the eos token in the completion
+                            content = content[: -len(eos_str)]
+            except Exception as e:
+                logger.exception(f"Failed to parse llm response: {r}")
+                raise e
+            output = LLMOutput(content=content)
+            llm_call = self.log_output(prompts[i], output)
+            llm_call.logprobs = logprobs
+            result.append(llm_call)
+        self._stats["time_postprocess_llm_response"].append(time.time() - start_postprocess_time)
+        return result
```

Comment:
Side note: class `TrainableLLM` starts looking really messy. It's really about logprobs computation and not LM training, and I can see some duplicated code. But we can definitely fix it later in a separate PR.